jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/BPMN2FileIndexer.java

```diff
@@ -0,0 +1,162 @@
+
+        try {
+            ProcessDefinition procesDef = getProcessDefinition(bpmnStr, path);
+            if( procesDef != null) {
+                ProcessDefintionIndexVisitor visitor = new ProcessDefintionIndexVisitor(procesDef, builder);
+                visitor.visit();
+                addReferencedResourcesToIndexBuilder(builder, visitor);
+            }
+        } catch( Exception e ) {
+            // log and ignore
+            logger.info("Indexing hampered because BPMN2 compilation failed [" + path.toString() + "]: " + e.getMessage() );
+        }
+
+        /**
+         * IMPORTANT: sometimes -- just sometimes, we might be indexing an INCORRECT BPMN2! (How dare they! ;D )
+         *
+         * Well, in that case, while we normally would be able to get everything from the ProcessDefinition,
+         * we can't -- because the Bpmn2DataServiceImpl will go BOOM and we will have nothing.
+         *
+         * Sooo.. that's why we *also* use the ProcessIndexVisitor! It quickly goes through the node-tree in the Process
+         * and grabs what it can..
+         */
+
+        ClassLoader projectClassLoader = getProjectClassLoader(project);
+        List<Process> processes = getProcesses(bpmnStr, projectClassLoader );
+        if( processes != null ) {
+            for( Process process : processes ) {
+                Resource resource = new ReaderResource( new StringReader( bpmnStr ) );
+                ProcessValidationError[] errors;
+
+                ProcessValidator validator = ProcessValidatorRegistry.getInstance().getValidator( process, resource );
+                errors = validator.validateProcess( process );
+                if ( errors.length > 0 ) {
+                    logger.error( "The process: " + process.getName() + " has validation errors, but we can still try to index information for well defined nodes." );
+                }
+
+                ProcessIndexVisitor visitor = new ProcessIndexVisitor( process, builder );
+                visitor.visit();
+
+                addReferencedResourcesToIndexBuilder(builder, visitor);
+            }
+        } else {
+            logger.warn( "No process was found in file: " + path.toUri() );
+        }
+
+        return builder;
+    }
+
+    private ProcessDefinition getProcessDefinition(String bpmnStr, Path path) throws SAXException, IOException {
+        final KieProject project = projectService.resolveProject( Paths.convert( path ) );
+        String deploymentId = project.getPom().getGav().toString();
+
+        return bpmn2Service.buildProcessDefinition(deploymentId, bpmnStr, null, false);
+    }
+
+    private List<Process> getProcesses(String bpmnStr, ClassLoader projectClassLoader) throws SAXException, IOException {
+        XmlProcessReader processReader = new XmlProcessReader( modules, projectClassLoader );
+        return processReader.read( new StringReader(bpmnStr) );
+    }
+
+    @Override
+    public KObjectKey toKObjectKey( Path path ) {
```

Comment:
I thought I read somewhere that this method should not be overridden and ideally it'd have been ```final```?